win32ss/gdi/ntgdi/freetype.c

```diff
@@ -715,6 +711,83 @@ InitFontSupport(VOID)
     return TRUE;
 }
 
+static VOID
+IntFreeFontCacheList(PLIST_ENTRY pHead)
+{
+    PLIST_ENTRY pEntry, pNextEntry;
+    PFONT_CACHE_ENTRY pFontCache;
+
+    ASSERT_FREETYPE_LOCK_HELD();
+
+    for (pEntry = pHead->Flink; pEntry != pHead; pEntry = pNextEntry)
+    {
+        pNextEntry = pEntry->Flink;
+        pFontCache = CONTAINING_RECORD(pEntry, FONT_CACHE_ENTRY, ListEntry);
+        RemoveCachedEntry(pFontCache);
+    }
+
+    pHead->Flink = pHead;
```

Comment:
A to-the-point and correct implementation, that also follows existing usage, would be this:
```suggestion
    while (!IsListEmpty(pHead))
    {
        pEntry = RemoveHeadList(pHead);
        pFontCache = CONTAINING_RECORD(pEntry, FONT_CACHE_ENTRY, ListEntry);
        RemoveCachedEntry(pFontCache);
    }
```
This would ensure:
- that the entry gets removed from the list while maintaining correct list iteration;
- automatically ensure that at the end, the linked-list is still correctly set up (with its pointers) as an empty list.

Reply:
OK, but shouldn't each list entry be removed in `CleanupFontEntryEx()`, as it was suggested in https://github.com/reactos/reactos/pull/6955#discussion_r1616368605? :thinking:

Reply:
Fixed.